plugins/genesys/fu-genesys-usbhub-firmware.c

```diff
@@ -14,10 +14,55 @@
 struct _FuGenesysUsbhubFirmware {
 	FuFirmwareClass parent_instance;
 	FuGenesysStaticToolString static_ts;
+	FuGenesysChip chip;
 };
 
 G_DEFINE_TYPE(FuGenesysUsbhubFirmware, fu_genesys_usbhub_firmware, FU_TYPE_FIRMWARE)
 
+static gboolean
+fu_genesys_usbhub_firmware_get_chip(FuGenesysUsbhubFirmware *self,
+				    const guint8 *buf,
+				    gsize bufsz,
+				    GError **error)
+{
+	guint8 project_ic_type[6];
+
+	if (!fu_memcpy_safe(project_ic_type,
+			    sizeof(project_ic_type),
+			    0, /* dst */
+			    buf,
+			    bufsz,
+			    GENESYS_USBHUB_STATIC_TOOL_STRING_OFFSET_GL3523 + 8, /* src */
+			    sizeof(project_ic_type),
+			    error))
+		return FALSE;
+
+	if (memcmp(project_ic_type, "3523", 4) == 0) {
+		self->chip.model = ISP_MODEL_HUB_GL3523;
+		self->chip.revision = 10 * (project_ic_type[4] - '0') + (project_ic_type[5] - '0');
+		return TRUE;
+	}
+
+	if (!fu_memcpy_safe(project_ic_type,
+			    sizeof(project_ic_type),
+			    0, /* dst */
+			    buf,
+			    bufsz,
+			    GENESYS_USBHUB_STATIC_TOOL_STRING_OFFSET_GL3590 + 8, /* src */
+			    sizeof(project_ic_type),
+			    error))
+		return FALSE;
+
+	if (memcmp(project_ic_type, "3590", 4) == 0) {
+		self->chip.model = ISP_MODEL_HUB_GL3590;
+		self->chip.revision = 10 * (project_ic_type[4] - '0') + (project_ic_type[5] - '0');
+		return TRUE;
+	}
+
+	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "Unsupported IC");
```

Comment:
Can we add the following?

```
g_autofree gchar *ic_type =                             
fu_common_strsafe((const gchar *)mask_project_ic_type, sizeof(mask_project_ic_type));
g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "IC type %s not supported", ic_type); 
```

Reply:
The field may not contain correct string in "IC type". There are different offsets of firmware to indicate IC type.

Reply:
oh yes I remember why I have not add it :)